batchflow/models/torch/base.py

```diff
@@ -942,16 +954,54 @@ def train(self, *args, feed_dict=None, fetches=None, use_lock=True, profile=Fals
             callback.on_iter_end()
         return output
 
-    def _train(self, *args, fetches=None, sync_frequency=True):
+    def _train(self, *args, fetches=None, sync_frequency=True, sam_parameter_placeholder=False, sam_rho=0.05):
         # Parse inputs
         *inputs, targets = args
         inputs = inputs[0] if isinstance(inputs, (tuple, list)) and len(inputs) == 1 else inputs
 
         # Apply model, compute loss and gradients
         predictions = self.model(inputs)
+
+        if self.iteration >= 1 and sam_parameter_placeholder:
+            # Store grads from previous microbatches
+            for p in self.model.parameters():
+                if p.grad is not None:
+                    self.optimizer.state[p]['previous_grad'] = p.grad.clone().detach()
+                    p.grad = None
+
         loss = sum(loss_fn(predictions, targets) for loss_fn in self.loss) / len(self.loss)
         loss.backward()
 
+        if self.iteration >= 1 and sam_parameter_placeholder:
+            #pylint: disable=protected-access
+            # Use obtained grad to move to the local maxima
+            grads = []
+            params_with_grads = []
+            for p in self.model.parameters():
+                if p.grad is not None:
+                    grads.append(p.grad.clone().detach())
+                    params_with_grads.append(p)
+                    p.grad = None
+            grad_norm = torch.stack([g.detach().norm(2).to(self.device) for g in grads]).norm(2)
+            epsilons = [eps * sam_rho / grad_norm for eps in grads]
```

Comment:
What about making `grads` a tensor and scaling it 'simultaneously', instead of iterating over a list?